File: android-app/app/src/main/java/com/lasohealth/android/core/ml/MLOrchestrator.kt
package com.lasohealth.android.core.ml

import android.util.Log
import com.lasohealth.android.core.analysis.DailySample
import com.lasohealth.android.core.analysis.MetricBaseline
import com.lasohealth.android.core.model.HealthMetric
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Orchestrates all on-device ML components: feature extraction, forecasting,
 * correlation discovery, anomaly detection, pattern mining, health state
 * classification, predictive scoring, health signals, and decision policy.
 *
 * Call [runPipeline] from a coroutine scope (e.g. a ViewModel) after fresh
 * health data has been synced. The orchestrator runs each component in
 * sequence on [Dispatchers.Default] and publishes results as plain properties
 * that the UI layer can observe.
 *
 * When fewer than 7 days of data are available, population-based defaults are
 * returned (cold-start).
 *
 * Ported from iOS `MLOrchestrator.swift`.
 */
class MLOrchestrator {

    companion object {
        private const val TAG = "MLOrchestrator"
    }

    // -- ML Components (Core) ---------------------------------------------------

    val featureEngine = FeatureEngine()
    val forecaster = TimeSeriesForecaster()
    val correlationDiscovery = CorrelationDiscovery()
    val anomalyDetector = AdaptiveAnomalyDetector()
    val patternMiner = PatternMiner()
    val stateClassifier = HealthStateClassifier()
    val predictiveScorer = PredictiveScorer()

    // -- ML Components (Intelligence Layer) -------------------------------------

    val predictiveHealthSignals = PredictiveHealthSignals()
    val decisionPolicyEngine = DecisionPolicyEngine()
    val conformalCalibrator = ConformalCalibrator()

    // -- Observable state -------------------------------------------------------

    /** Whether the pipeline has been run at least once. */
    var hasRunOnce: Boolean = false
        private set

    /** Whether the pipeline is currently executing. */
    var isRunning: Boolean = false
        private set

    /** Personalization stage based on data maturity. */
    var personalizationStage: PersonalizationStage = PersonalizationStage.POPULATION
        private set

    /** Personalization status for the UI. */
    var personalizationStatus: PersonalizationBlender.PersonalizationStatus? = null
        private set

    /** Data sufficiency assessment. */
    var dataSufficiency: UncertaintyEstimator.DataSufficiency? = null
        private set

    /** Component readiness. */
    var componentReadiness: List<UncertaintyEstimator.ComponentReadiness> = emptyList()
        private set

    /** True when any component has enough data to produce results. */
    val isReady: Boolean
        get() = featureEngine.isReady || forecaster.isReady || correlationDiscovery.isReady

    // -- Results ----------------------------------------------------------------

    /** Multi-horizon forecasts keyed by metric. */
    var predictions: Map<HealthMetric, MultiHorizonForecast> = emptyMap()
        private set

    /** Discovered metric correlations (FDR-corrected). */
    var correlations: List<MLCorrelation> = emptyList()
        private set

    /** Feature vectors from the last pipeline run (oldest first). */
    var featureVectors: List<DailyFeatureVector> = emptyList()
        private set

    /** Discovered periodic patterns. */
    var patterns: List<DiscoveredPattern> = emptyList()
        private set

    /** All identified health states. */
    var healthStates: List<HealthState> = emptyList()
        private set

    /** Current health state. */
    var currentHealthState: HealthState? = null
        private set

    /** HMM-smoothed state history. */
    var smoothedStates: List<SmoothedHealthState> = emptyList()
        private set

    /** Adaptive anomaly results. */
    var adaptiveAnomalies: List<AdaptiveAnomalyDetector.AdaptiveAnomaly> = emptyList()
        private set

    /** Tomorrow risk prediction. */
    var tomorrowRiskPrediction: MLPrediction? = null
        private set

    /** Predictive health signal report. */
    var healthSignalReport: PredictiveHealthSignals.HealthSignalReport? = null
        private set

    /** Today's action recommendation. */
    var policyDecision: PolicyDecision? = null
        private set

    // -- Pipeline ---------------------------------------------------------------

    /**
     * Run the full ML pipeline sequentially.
     *
     * Pipeline stages:
     * 1. Feature extraction
     * 2. Time-series forecasting (train + forecast)
     * 3. Correlation discovery
     * 4. Anomaly detection (Isolation Forest)
     * 5. Pattern mining (autocorrelation + Goertzel)
     * 6. Health state classification (GMM + HMM)
     * 7. Predictive scoring (GBT)
     * 8. Predictive health signals (6 detectors)
     * 9. Decision policy engine (expected utility)
     * 10. Uncertainty estimation + personalization status
     *
     * All computation runs on [Dispatchers.Default].
     */
    suspend fun runPipeline(
        timeSeries: Map<HealthMetric, List<DailySample>>,
        baselines: Map<HealthMetric, MetricBaseline>,
        scoreHistory: List<Pair<Long, Int>> = emptyList(),
        anomalyCounts: Map<Long, Int> = emptyMap(),
    ) = withContext(Dispatchers.Default) {
        if (isRunning) return@withContext
        isRunning = true
        val startTime = System.currentTimeMillis()

        try {
            val dayCount = computeDayCount(timeSeries)
            personalizationStage = PersonalizationStage.from(dayCount)

            Log.d(TAG, "Pipeline start: $dayCount days of data, stage=$personalizationStage")

            // Update uncertainty/readiness regardless of data maturity.
            dataSufficiency = UncertaintyEstimator.assessDataSufficiency(
                totalDays = dayCount,
                metricsPresent = timeSeries.size,
                requiredMetrics = 10,
            )
            componentReadiness = UncertaintyEstimator.checkComponentReadiness(dayCount)
            personalizationStatus = PersonalizationBlender.personalizationStatus(
                userDataDays = dayCount,
                metricsTracked = timeSeries.size,
            )

            // Cold-start: population-based defaults.
            if (dayCount < FeatureEngine.MINIMUM_DAYS) {
                applyColdStartDefaults(timeSeries)
                return@withContext
            }

            // Step 1: Feature extraction.
            val allSamples = timeSeries.values.flatten()
            val vectors = featureEngine.extract(allSamples, baselines)
            featureVectors = vectors
            Log.d(TAG, "Features: ${vectors.size} vectors, ${featureEngine.orderedKeys.size} keys")

            // Step 2: Forecasting.
            forecaster.train(timeSeries)
            val forecasts = mutableMapOf<HealthMetric, MultiHorizonForecast>()
            for (metric in timeSeries.keys) {
                forecaster.forecast(metric, horizons = listOf(1, 3, 7))?.let { forecasts[metric] = it }
            }
            predictions = forecasts
            Log.d(TAG, "Forecasts: ${forecasts.size} metrics")

            // Step 3: Correlation discovery.
            if (dayCount >= CorrelationDiscovery.MINIMUM_DAYS) {
                correlationDiscovery.discover(timeSeries)
                correlations = correlationDiscovery.correlations
                Log.d(TAG, "Correlations: ${correlations.size} discovered, ${correlationDiscovery.fdrSignificant().size} FDR-significant")
            }

            // Step 4: Anomaly detection.
            if (dayCount >= AdaptiveAnomalyDetector.MINIMUM_DAYS && vectors.isNotEmpty()) {
                anomalyDetector.train(vectors, featureEngine.orderedKeys, baselines)
                adaptiveAnomalies = anomalyDetector.scoreRecent(vectors)
                val surfaced = adaptiveAnomalies.count { it.shouldSurface }
                Log.d(TAG, "Anomalies: ${adaptiveAnomalies.size} scored, $surfaced surfaced")
            }

            // Step 5: Pattern mining.
            if (dayCount >= PatternMiner.MINIMUM_DAYS) {
                patternMiner.mine(timeSeries)
                patterns = patternMiner.patterns
                Log.d(TAG, "Patterns: ${patterns.size} discovered")
            }

            // Step 6: Health state classification.
            if (dayCount >= HealthStateClassifier.MINIMUM_DAYS && vectors.isNotEmpty()) {
                stateClassifier.classify(vectors, featureEngine.orderedKeys)
                healthStates = stateClassifier.states
                currentHealthState = stateClassifier.currentState
                smoothedStates = stateClassifier.smoothedStates(vectors)
                Log.d(TAG, "States: ${healthStates.size} identified, current=${currentHealthState?.label}")
            }

            // Step 7: Predictive scoring (GBT).
            if (dayCount >= PredictiveScorer.MINIMUM_DAYS && vectors.isNotEmpty() && scoreHistory.isNotEmpty()) {
                predictiveScorer.train(vectors, featureEngine.orderedKeys, scoreHistory, anomalyCounts)
                if (predictiveScorer.isReady) {
                    tomorrowRiskPrediction = predictiveScorer.predict(vectors.last())
                    Log.d(TAG, "Risk prediction: ${tomorrowRiskPrediction?.predictedValue}")
                }
            }

            // Step 8: Predictive health signals.
            if (dayCount >= PredictiveHealthSignals.MINIMUM_DAYS) {
                healthSignalReport = predictiveHealthSignals.analyze(timeSeries, baselines)
                Log.d(TAG, "Health signals: ${healthSignalReport?.signals?.size ?: 0} signals, " +
                    "risk=${healthSignalReport?.overallRisk}")
            }

            // Step 9: Decision policy engine.
            policyDecision = decisionPolicyEngine.decide(
                forecasts = forecasts,
                baselines = baselines,
                anomalies = adaptiveAnomalies,
                healthSignals = healthSignalReport,
                patterns = patterns,
                correlations = correlationDiscovery.fdrSignificant(),
                currentState = currentHealthState,
                dayCount = dayCount,
            )
            Log.d(TAG, "Policy: ${policyDecision?.action}")

        } catch (e: Exception) {
            Log.e(TAG, "Pipeline failed", e)
        } finally {
            isRunning = false
            hasRunOnce = true
            val elapsed = System.currentTimeMillis() - startTime
            Log.d(TAG, "Pipeline complete in ${elapsed}ms")
        }
    }

    // -- Cold-start -------------------------------------------------------------

    private fun applyColdStartDefaults(timeSeries: Map<HealthMetric, List<DailySample>>) {
        predictions = emptyMap()
        correlations = emptyList()
        featureVectors = emptyList()
        patterns = emptyList()
        healthStates = emptyList()
        currentHealthState = null
        smoothedStates = emptyList()
        adaptiveAnomalies = emptyList()
        tomorrowRiskPrediction = null
        healthSignalReport = null

        // Cold-start prediction from population priors.
        val availableData = timeSeries.mapValues { (_, samples) ->
            samples.sortedBy { it.date }.map { it.value }
        }
        val (riskScore, confidence, _) = PersonalizationBlender.coldStartPrediction(availableData)

        policyDecision = if (riskScore > 0.5) {
            PolicyDecision(
                action = "Take it easy today",
                detail = "Based on general health benchmarks, your metrics suggest taking a lighter approach today.",
                confidence = confidence,
                metric = null,
                iconEmoji = "\uD83D\uDCA4",
            )
        } else {
            PolicyDecision(
                action = "Keep tracking",
                detail = "We need a few more days of data to personalise insights for you.",
                confidence = 0.3,
                metric = null,
                iconEmoji = "\uD83D\uDCCA",
            )
        }
    }

    // -- Helpers ----------------------------------------------------------------

    /**
     * Compute the number of distinct calendar days across all metric series.
     */
    private fun computeDayCount(timeSeries: Map<HealthMetric, List<DailySample>>): Int {
        val allDates = timeSeries.values.flatMap { samples -> samples.map { startOfDay(it.date) } }.toSet()
        return allDates.size
    }

    private fun startOfDay(epochMillis: Long): Long {
        val cal = java.util.Calendar.getInstance()
        cal.timeInMillis = epochMillis
        cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
        cal.set(java.util.Calendar.MINUTE, 0)
        cal.set(java.util.Calendar.SECOND, 0)
        cal.set(java.util.Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }
}

File: Core/Analysis/ML/MLOrchestrator.swift
import Foundation
import Observation
import os

/// Manages all ML component lifecycle, sequential execution, incremental training, and periodic retraining.
@Observable
final class MLOrchestrator {

    private let logger = Logger(subsystem: "com.healthpulse.ml", category: "MLOrchestrator")

    // MARK: - ML Components (Core)

    let featureEngine = FeatureEngine()
    let forecaster = TimeSeriesForecaster()
    let anomalyDetector = AdaptiveAnomalyDetector()
    let patternMiner = PatternMiner()
    let correlationDiscovery = CorrelationDiscovery()
    let stateClassifier = HealthStateClassifier()
    let predictiveScorer = PredictiveScorer()
    let adherenceTracker = AdherenceTracker()
    let circadianAnalyzer = CircadianAnalyzer()

    // MARK: - ML Components (New)

    let decisionPolicyEngine = DecisionPolicyEngine()
    let mlEvaluator = MLEvaluator()
    let conformalCalibrator = ConformalCalibrator()
    let receptivityEstimator = ReceptivityEstimator()
    let healthDataQueryEngine = HealthDataQueryEngine()

    // MARK: - ML Components (Intelligence Layer)

    let interactionEngine = InteractionEffectEngine()
    let temporalMiner = TemporalSequenceMiner()
    let changePointDetector = ChangePointDetector()
    let personalOptimizer = PersonalOptimizer()
    let compoundInsightEngine = CompoundInsightEngine()

    // MARK: - Delegates

    private let pipelineRunner = MLPipelineRunner()
    private let resultAggregator = MLResultAggregator()
    private let calibrationManager = MLCalibrationManager()

    // MARK: - State

    /// Whether ML analysis has been run at least once
    var hasRunOnce = false
    /// Whether ML is currently running
    var isRunning = false
    /// Last full retrain date
    var lastFullRetrain: Date?

    /// Last ML pipeline completion time. used for 1-hour TTL
    private var lastPipelineCompletion: Date?
    private static let pipelineTTL: TimeInterval = 3600  // 1 hour

    // MARK: - Results (for AnalysisEngine to consume)

    /// ML-powered anomalies (from forecaster + isolation forest)
    var mlAnomalies: [TimeSeriesForecaster.ForecastAnomaly] = []
    /// Adaptive anomaly results
    var adaptiveAnomalies: [AdaptiveAnomalyDetector.AdaptiveAnomaly] = []
    /// Discovered patterns
    var discoveredPatterns: [DiscoveredPattern] = []
    /// ML-discovered correlations
    var mlCorrelations: [MLCorrelation] = []
    /// Current health state
    var currentHealthState: HealthState?
    /// Tomorrow risk prediction
    var tomorrowRiskPrediction: MLPrediction?

    // MARK: - New ML Results

    /// Predictive health signal report (fatigue, burnout, overtraining, insomnia, immune, inactivity)
    var healthSignalReport: PredictiveHealthSignals.HealthSignalReport?
    /// Policy-based recommendation decision. single source of truth for daily action
    var policyDecision: PolicyDecision?
    /// Personalization status for the user
    var personalizationStatus: PersonalizationBlender.PersonalizationStatus?
    /// Data sufficiency assessment
    var dataSufficiency: UncertaintyEstimator.DataSufficiency?
    /// Component readiness
    var componentReadiness: [UncertaintyEstimator.ComponentReadiness] = []
    /// HMM-smoothed state history
    var smoothedStates: [SmoothedHealthState] = []
    /// Multi-horizon forecasts (1d, 3d, 7d) per metric
    var multiHorizonForecasts: [HealthMetric: TimeSeriesForecaster.MultiHorizonForecast] = [:]
    /// Multivariate regression results (omitted-variable-reduced causality)
    var multivariateResults: [MultivariateRegressionResult] = []
    /// Latest evaluation summary per component
    var evaluationSummaries: [String: ComponentEvaluation] = [:]
    /// Model drift alerts
    var driftAlerts: [DriftReport] = []
    /// Interaction effects (dose-response, conditional, moderation)
    var interactionEffects: [InteractionEffectEngine.InteractionEffect] = []
    /// Dose-response curves
    var doseResponseCurves: [InteractionEffectEngine.DoseResponseCurve] = []
    /// Temporal sequences (multi-step causal chains)
    var temporalSequences: [TemporalSequenceMiner.TemporalSequence] = []
    /// Precursor warning patterns
    var precursorPatterns: [TemporalSequenceMiner.PrecursorPattern] = []
    /// Compounding effects
    var compoundingEffects: [TemporalSequenceMiner.CompoundingEffect] = []
    /// Detected changepoints (regime shifts)
    var changePoints: [ChangePointDetector.ChangePoint] = []
    /// Regime comparisons per metric
    var regimeComparisons: [ChangePointDetector.RegimeComparison] = []
    /// Personal optimal profile
    var optimalProfile: PersonalOptimizer.OptimalProfile?
    /// Ideal day targets
    var idealDay: PersonalOptimizer.IdealDay?
    /// Score sensitivities per metric
    var scoreSensitivities: [PersonalOptimizer.SensitivityResult] = []
    /// Compound insights (cross-component synthesis)
    var compoundInsights: [CompoundInsightEngine.CompoundInsight] = []

    // MARK: - Cached Feature Vectors

    private let vectorLock = NSLock()
    private var _cachedVectors: [DailyFeatureVector] = []
    private var cachedVectors: [DailyFeatureVector] {
        get { vectorLock.withLock { _cachedVectors } }
        set { vectorLock.withLock { _cachedVectors = newValue } }
    }

    /// Latest feature vector (for simulation engine)
    var latestVector: DailyFeatureVector? { cachedVectors.last }

    /// Full state history from classifier (for Health State Timeline)
    var stateHistory: [(date: Date, label: String)] {
        stateClassifier.isReady ? stateClassifier.stateHistory : []
    }

    /// Transition matrix from classifier (for Health State Timeline)
    var stateTransitionMatrix: [String: [String: Double]] {
        stateClassifier.isReady ? stateClassifier.transitionMatrix : [:]
    }

    /// All identified health states from classifier
    var healthStates: [HealthState] {
        stateClassifier.isReady ? stateClassifier.states : []
    }

    /// Circadian profile (chronotype, peak times)
    var circadianProfile: CircadianAnalyzer.CircadianProfile? { circadianAnalyzer.profile }

    /// Timing recommendations from circadian analysis
    var timingRecommendations: [CircadianAnalyzer.TimingRecommendation] { circadianAnalyzer.recommendations }

    /// Whether circadian analysis should run
    var needsCircadianAnalysis: Bool { circadianAnalyzer.needsAnalysis }

    // MARK: - Run ML Analysis

    /// Run all ML components that are ready. Called from AnalysisEngine after rule-based analysis.
    /// Components run sequentially (one at a time) to avoid CPU spikes, ordered by priority.
    /// Thermal state is checked before starting and between each component.
    func runMLAnalysis(
        timeSeries: [HealthMetric: MetricTimeSeries],
        baselines: [HealthMetric: UserBaseline],
        trends: [HealthMetric: TrendAnalyzer.TrendResult],
        ruleBasedAnomalies: [AnomalyDetector.AnomalyResult],
        scoreHistory: [(date: Date, score: Int)],
        anomalyCounts: [Date: Int],
        focusCategories: Set<HealthCategory> = []
    ) async {
        // Remote kill switch checked here, at the single choke point, so every
        // entry path (dashboard refresh, background refresh, future callers)
        // respects it without per-caller guards.
        guard !RemoteConfigManager.shared.killMLPipeline else {
            logger.warning("Skipping ML analysis: kill_ml_pipeline is enabled")
            return
        }

        // Bail out entirely if device is already under thermal pressure
        let thermalState = ProcessInfo.processInfo.thermalState
        if thermalState == .critical || thermalState == .serious {
            logger.warning("Skipping ML analysis: thermal state is \(thermalState == .critical ? "critical" : "serious"). deferring to next cycle")
            return
        }

        // Skip if pipeline ran recently and no new data indicator
        if let lastCompletion = lastPipelineCompletion,
           Date().timeIntervalSince(lastCompletion) < Self.pipelineTTL,
           hasRunOnce {
            logger.info("Skipping ML analysis. pipeline ran \(Int(Date().timeIntervalSince(lastCompletion)))s ago (TTL: \(Int(Self.pipelineTTL))s)")
            return
        }

        guard !isRunning else {
            logger.info("Skipping ML analysis: pipeline already running")
            return
        }

        isRunning = true
        let pipelineStart = Date()
        var componentsRunCount = 0
        defer {
            isRunning = false
            hasRunOnce = true
            lastPipelineCompletion = Date()
            // Record completion time for BackgroundRefreshCoordinator incremental skip logic
            UserDefaults.standard.set(Date(), forKey: "lastMLPipelineCompletion")
            if componentsRunCount > 0 {
                let pipelineDurationMs = Int(Date().timeIntervalSince(pipelineStart) * 1000)
                let totalDataPoints = timeSeries.values.reduce(0) { $0 + $1.samples.count }
                Task { @MainActor in
                    AppAnalytics.shared.trackMLAnalysisPerformance(
                        durationMs: pipelineDurationMs,
                        componentsRun: componentsRunCount,
                        dataPointsUsed: totalDataPoints
                    )
                }
            }
        }

        let input = MLPipelineRunner.PipelineInput(
            timeSeries: timeSeries,
            baselines: baselines,
            trends: trends,
            ruleBasedAnomalies: ruleBasedAnomalies,
            scoreHistory: scoreHistory,
            anomalyCounts: anomalyCounts,
            focusCategories: focusCategories
        )

        let components = makeComponents()

        guard let pipelineOutput = await pipelineRunner.run(
            input: input,
            components: components,
            needsFullRetrain: needsFullRetrain,
            cachedVectorsSetter: { [self] vectors in cachedVectors = vectors }
        ) else {
            return
        }
        componentsRunCount = pipelineOutput.componentsRunCount

        let vectors = pipelineOutput.vectors
        guard !vectors.isEmpty else { return }

        // Apply pipeline output to observable state
        applyPipelineOutput(pipelineOutput)

        // Clear bulky intermediates and dead results to reduce memory pressure.
        // These properties are computed but never consumed by any ViewModel or View:
        multivariateResults = []  // Granger multivariate: O(N³) results, no consumer
        driftAlerts = []          // Welch's t-test drift: no consumer
        regimeComparisons = []    // ChangePoint regimes: no consumer
        compoundingEffects = []   // TemporalMiner compounding: intentionally disabled

        if !pipelineOutput.stoppedEarly {
            // Step 13: CompoundInsightEngine (uses current self.* state from previous run)
            compoundInsights = resultAggregator.synthesizeCompoundInsights(
                timeSeries: timeSeries,
                baselines: baselines,
                currentAdaptiveAnomalies: adaptiveAnomalies,
                currentMLCorrelations: mlCorrelations,
                currentDiscoveredPatterns: discoveredPatterns,
                currentTomorrowRiskPrediction: tomorrowRiskPrediction,
                smoothedStates: smoothedStates,
                scoreHistory: scoreHistory,
                components: components
            )
        }

        // Collect all results and compute policy decision
        let aggregated = resultAggregator.collectResults(
            timeSeries: timeSeries,
            vectors: vectors,
            baselines: baselines,
            trends: trends,
            ruleBasedAnomalies: ruleBasedAnomalies,
            scoreHistory: scoreHistory,
            focusCategories: focusCategories,
            components: components,
            calibrationManager: calibrationManager,
            evaluationSummaries: evaluationSummaries,
            dataSufficiency: dataSufficiency,
            smoothedStates: smoothedStates,
            timingRecommendations: timingRecommendations
        )
        applyAggregatedResults(aggregated)
    }

    // MARK: - Incremental Training

    /// Lightweight daily update without full retrain
    func trainIncremental(
        timeSeries: [HealthMetric: MetricTimeSeries],
        baselines: [HealthMetric: UserBaseline],
        todayScore: Int,
        todayAnomalyCount: Int
    ) {
        pipelineRunner.trainIncremental(
            timeSeries: timeSeries,
            baselines: baselines,
            todayScore: todayScore,
            todayAnomalyCount: todayAnomalyCount,
            components: makeComponents(),
            cachedVectors: cachedVectors
        )
    }

    // MARK: - Full Retrain

    /// Full retrain of all components (expensive, every 30 days)
    var needsFullRetrain: Bool {
        guard let lastRetrain = lastFullRetrain else { return hasRunOnce }
        return Date().timeIntervalSince(lastRetrain) > 30 * 24 * 3600
    }

    func markFullRetrainComplete() {
        lastFullRetrain = Date()
    }

    /// Invalidate pipeline TTL so the next run proceeds regardless
    func invalidatePipelineTTL() {
        lastPipelineCompletion = nil
    }

    // MARK: - Circadian Analysis (separate pipeline, runs weekly)

    /// Run circadian analysis using hourly data. Called separately from the main ML pipeline.
    func runCircadianAnalysis(hourlyData: [HealthMetric: [[Double]]]) {
        // Same choke-point kill switch as the main pipeline above.
        guard !RemoteConfigManager.shared.killMLPipeline else {
            logger.warning("Skipping circadian analysis: kill_ml_pipeline is enabled")
            return
        }
        guard !hourlyData.isEmpty else { return }
        circadianAnalyzer.analyze(hourlyData: hourlyData)
    }

    // MARK: - Insight Generation

    /// Convert ML results into Insight objects for the insights pipeline
    func generateInsights() -> [Insight] {
        resultAggregator.generateInsights(
            discoveredPatterns: discoveredPatterns,
            currentHealthState: currentHealthState,
            tomorrowRiskPrediction: tomorrowRiskPrediction,
            healthSignalReport: healthSignalReport,
            interactionEffects: interactionEffects,
            precursorPatterns: precursorPatterns,
            temporalSequences: temporalSequences,
            changePoints: changePoints,
            optimalProfile: optimalProfile,
            idealDay: idealDay,
            mlCorrelations: mlCorrelations,
            compoundInsightEngine: compoundInsightEngine,
            circadianAnalyzer: circadianAnalyzer,
            cachedVectorCount: cachedVectors.count
        )
    }

    // MARK: - Private Helpers

    private func makeComponents() -> MLPipelineRunner.Components {
        MLPipelineRunner.Components(
            featureEngine: featureEngine,
            forecaster: forecaster,
            anomalyDetector: anomalyDetector,
            patternMiner: patternMiner,
            correlationDiscovery: correlationDiscovery,
            stateClassifier: stateClassifier,
            predictiveScorer: predictiveScorer,
            interactionEngine: interactionEngine,
            temporalMiner: temporalMiner,
            changePointDetector: changePointDetector,
            personalOptimizer: personalOptimizer,
            compoundInsightEngine: compoundInsightEngine,
            mlEvaluator: mlEvaluator,
            conformalCalibrator: conformalCalibrator,
            decisionPolicyEngine: decisionPolicyEngine
        )
    }

    private func applyPipelineOutput(_ output: MLPipelineRunner.PipelineOutput) {
        dataSufficiency = output.dataSufficiency
        componentReadiness = output.componentReadiness
        personalizationStatus = output.personalizationStatus
        smoothedStates = output.smoothedStates
        multiHorizonForecasts = output.multiHorizonForecasts
        multivariateResults = output.multivariateResults
        healthSignalReport = output.healthSignalReport
        driftAlerts = output.driftAlerts
        evaluationSummaries = output.evaluationSummaries
        interactionEffects = output.interactionEffects
        doseResponseCurves = output.doseResponseCurves
        temporalSequences = output.temporalSequences
        precursorPatterns = output.precursorPatterns
        compoundingEffects = output.compoundingEffects
        changePoints = output.changePoints
        regimeComparisons = output.regimeComparisons
        optimalProfile = output.optimalProfile
        idealDay = output.idealDay
        scoreSensitivities = output.scoreSensitivities
    }

    private func applyAggregatedResults(_ results: MLResultAggregator.AggregatedResults) {
        mlAnomalies = results.mlAnomalies
        adaptiveAnomalies = results.adaptiveAnomalies
        discoveredPatterns = results.discoveredPatterns
        mlCorrelations = results.mlCorrelations
        currentHealthState = results.currentHealthState
        tomorrowRiskPrediction = results.tomorrowRiskPrediction
        policyDecision = results.policyDecision
    }
}
